src/WorkspaceAction.ts

```diff
@@ -0,0 +1,151 @@
+import { getWorkspaces } from "coder/site/src/api/api"
+import { Workspace, WorkspacesResponse } from "coder/site/src/api/typesGenerated"
+import { formatDistanceToNowStrict } from "date-fns"
+import * as vscode from "vscode"
+
+interface NotifiedWorkspace {
+  workspace: Workspace
+  wasNotified: boolean
+  impendingActionDeadline: string
+}
+
+export class WorkspaceAction {
+  #fetchWorkspacesInterval?: ReturnType<typeof setInterval>
+
+  #ownedWorkspaces: Workspace[] = []
+  #workspacesApproachingAutostop: NotifiedWorkspace[] = []
+  #workspacesApproachingDeletion: NotifiedWorkspace[] = []
+
+  private constructor(private readonly vscodeProposed: typeof vscode, ownedWorkspaces: Workspace[]) {
+    this.#ownedWorkspaces = ownedWorkspaces
+
+    // seed initial lists
+    this.seedNotificationLists()
+
+    this.notifyAll()
+
+    // set up polling so we get current workspaces data
+    this.pollGetWorkspaces()
+  }
+
+  static async init(vscodeProposed: typeof vscode) {
+    // fetch all workspaces owned by the user and set initial public class fields
+    let ownedWorkspacesResponse: WorkspacesResponse
+    try {
+      ownedWorkspacesResponse = await getWorkspaces({ q: "owner:me" })
+    } catch (error) {
+      ownedWorkspacesResponse = { workspaces: [], count: 0 }
+    }
+    return new WorkspaceAction(vscodeProposed, ownedWorkspacesResponse.workspaces)
+  }
+
+  seedNotificationLists() {
```

Comment:
Very minor nit: `seed` implies to me we call it once but we call it on every poll; maybe `updateNotificationList` instead?

Reply:
One other thought on the notification list: if you get notified for a workspace being autostopped then you start it again leaving VS Code open, would you not get notified for the next autostop?  Maybe if the deadline changes we want to reset `wasNotified`.

Reply:
Ah I guess it would only be a problem if it autostops within an hour of you starting it again, otherwise the workspace would have been removed from the list and thus essentially resetting `wasNotified`.

Reply:
I had to think about this for a minute! I think you're saying that as we poll, we are redefining `workspacesApproachingAutostop`, effectively resetting `wasNotified`. If a workspace is temporarily removed from that list and then added back, `wasNotified` will be reinitialized to `true`. I agree - should only be an issue if the workspace is eligible to autostop within the hour, in which case I think the extra notification is forgivable, but let me know if you feel that will be annoying to the user.

Reply:
I was thinking about the opposite case where we might miss a notification because `wasNotified` was set for an earlier/different deadline and never cleared.

Specifically I was thinking about it happening when a workspace automatically turns off and then the user turns it on again but I just realized that would not happen even if the auto-stop was set to one hour because the job status and transition would no longer be `succeeded`/`start` and it would get kicked out of the list, resetting `wasNotified`, and all is well, unless they somehow manage to start it so fast the poll does not catch it in the off state.  :facepalm:  Sorry for the noise.

I suppose we could technically hit the issue if the deadline is extended (by one hour or less)? 

1. User gets notified that a workspace is about to auto-stop (or be deleted)
2. `wasNotified` set to true
3. User extends deadline (or delete deadline is moved) by one hour or less
4. No new notification for the next deadline since already notified

But since VS Code itself is supposed to keep the workspace active we would never see this case anyway?  And shifting a delete by one hour seems like a rather unlikely case as well.  So I think you can ignore everything I am saying. :sweat_smile: 

If we do need to account for this I think we could do it easily by only preserving `wasNotified` if `deadline`/`deleting_at` is the same.  For example:

```typescript
    const wasNotified =
      this.#workspacesApproachingAutostop.find((wn) => wn.workspace.id === workspace.id && wn.latest_build.deadline == workspace.latest_build.deadline)?.wasNotified ?? false
```

Reply:
Gotcha! I am going to be working on adding CTAs as a fast-follow so I will investigate extending auto-stop then!